hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -1847,24 +1849,49 @@ void sortDatanodes(String clientMachine, OmKeyInfo... keyInfos) {
     }
   }
 
-  private List<DatanodeDetails> sortDatanodes(String clientMachine,
+  public List<DatanodeDetails> sortDatanodes(String clientMachine,
       List<DatanodeDetails> nodes, OmKeyInfo keyInfo, List<String> nodeList) {
-    List<DatanodeDetails> sortedNodes = null;
-    try {
-      sortedNodes = scmClient.getBlockClient()
-          .sortDatanodes(nodeList, clientMachine);
-      if (LOG.isDebugEnabled()) {
-        LOG.debug("Sorted datanodes {} for client {}, result: {}", nodes,
-            clientMachine, sortedNodes);
+    DatanodeDetails client = null;
+    List<DatanodeDetails> possibleClients =
+        getClientNodesByAddress(clientMachine, nodes);
+    if (possibleClients.size() > 0) {
+      client = possibleClients.get(0);
+    } else {
+      /**
+       * In case of read from a non-datanode host, return
```

Comment:
What's the capability within SCM that cannot be replicated within Ozone for clients not on Datanodes?

Reply:
Good idea, I have made changes for resolving node location for reads from non-datanode hosts within `KeyManagerImpl` itself instead of making a call to the original `ScmBlockLocationProtocol#sortDatanodes` method.